Port PumpkinOS to ARM-based Macs

gernb opened the pull request:
This is a fully functional macOS-native app that builds in Xcode and creates a macOS "GUI" app. It is for Apple Silicon Macs (not Intel-based Macs; but maybe it will work there, too?).

gernb committed:
Initial stab at getting this working on macOS

gernb committed:
fix typo

gernb committed:
More fixes/updates

gernb committed:
dedicated macos configs

migueletto commented:
Hi, thanks for your interest in PumpkinOS. Given the nature of this pull request, I think the best way to proceed is to create a 'macos' branch and commit your changes there. Later, when it is stable, it would be merged into master. What do you think ?

gernb commented:
> Hi, thanks for your interest in PumpkinOS. Given the nature of this pull request, I think the best way to proceed is to create a 'macos' branch and commit your changes there. Later, when it is stable, it would be merged into master. What do you think ?

Sure, no problem. I have a working macOS-native (no SDL) version in my repo, but I didn’t think you’d want that merged into your repo since it made more changes to the core code to fix some bugs and to support data-isolation into the application’s support directories. Also, as a true macOS application, it builds from within Xcode and runs from a macOS app, rather than from the command line. Feel free to close this PR if you don’t want this SDL version for macOS.

gernb committed:
Native macOS application (#1)

Builds in Xcode, not via the command line. No dependency on SDL2.

gernb commented:
On second thought, I have merged my native macOS changes into my repo's master branch and made almost all my changes build for macOS (Darwin) only. Other than a couple of bug fixes and one change to Launcher (to not exit when the Home key is pressed), there should be no impact to Linux/Windows/etc.

This is fully functional on macOS, except for audio. I have not implemented any audio support (yet).

gernb commented on src/Launcher/Launcher.c:
```diff
@@ -2780,6 +2780,10 @@ static void EventLoop(launcher_data_t *data) {
   do {
     CheckNotifications();
     EvtGetEvent(&event, 30);
+    if ((event.eType == keyDownEvent) && (event.data.keyDown.chr == vchrLaunch) && (event.data.keyDown.modifiers & commandKeyMask)) {
```
This change prevents Launcher from quitting when the Home key is pressed.

gernb commented on src/libpit/main.c:
```diff
@@ -67,9 +67,9 @@ int pit_main(int argc, char *argv[], void (*callback)(int pe, void *data), void
-  debug_init(debugfile);
   ptr_init();
   thread_init();
+  debug_init(debugfile);
```
This fixes a crash when the log level is set to 2, since the logging subsystem requires the ptr and thread subsystems to be initialized first.

gernb commented on src/libpit/util.c:
```diff
@@ -7,7 +7,7 @@ char *getext(char *file) {
   ext = NULL;
 
-  for (i = sys_strlen(file)-1; i > 0; i--) {
+  for (i = sys_strlen(file)-1; i >= 0; i--) {
```
This fixes a crash when attempting to get the file suffix for a file the begins with a dot (such as `.empty` or `.DS_Store`.

gernb commented on src/mk.sh:
```diff
@@ -30,6 +30,9 @@ elif [ $OSNAME = "Serenity" ]; then
   fi
   SDL2=liblsdl2
   GUI=linux
+elif [ $OSNAME = "Darwin" ]; then
```
This allows the Xcode project to call this script to clean up build intermediates created by the makefiles.

gernb committed:
Include libshell and liboshell in app

gernb committed:
Add UI for filtering logs and module-specific logging

gernb committed:
Add support for installing apps via drag-n-drop

gernb committed:
Bug fix: Wrong Launcher window dimensions

When starting from a completely fresh state (no previous install/run), the Launcher requests a window size of 640x480 regardless of the actual display dimensions. So, if the display is, for example, 320x320 the Launcher window will be rendered incorrectly until after PumpkinOS is shutdown and restarted.

gernb committed:
Bug fix: Crash when shutting down after installing an app

When you install an app, the `pumpkin_deploy_file()` function is executed. This function always loads the fonts from `BOOT.prc` every time it runs, which is a problem because this causes the system fonts to get re-loaded on a different thread. So, when the system shuts down the font memory is free’d from the wrong thread heap, causing a crash.

The fix is to only load the `BOOT.prc` fonts after deploying files from libos.c.

migueletto commented:
I have merged your PR into the 'macos' branch.
